### PR TITLE
Using System Allocator to manage small objects used by Vulkan.

### PR DESCRIPTION
**Pull request by Jackie9527**
Signed-off-by: Jackie9527 <80555200+Jackie9527@users.noreply.github.com>

## What does this PR do?

Create a specific allocator to manage small objects used by Vulkan.

## How was this PR tested?

Open Sponza using vulkan as backend.
rhi-vulkan-memory-allocator

### CODE REVIEW

**Review by rgba16f — APPROVED** [review]

File: Gems/Atom/RHI/Vulkan/Code/Source/RHI/Allocator.h [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+    {
+        class VkSystemAllocator 
+        {
+
+        public:
+            static VkAllocationCallbacks* Get();
+        private:
+            VkSystemAllocator();
+            ~VkSystemAllocator();
+        private:
```
Comment (moraaar) [inline Gems/Atom/RHI/Vulkan/Code/Source/RHI/Allocator.h:26]: uber nit: private block repeated.

File: Gems/Atom/RHI/Vulkan/Code/Source/RHI/Allocator.h [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+#pragma once
+#include <AzCore/base.h>
+#include <AzCore/PlatformIncl.h>
+#include <vulkan/vulkan.h>
+
+namespace AZ
+{
+    namespace Vulkan
+    {
+        class VkSystemAllocator
```
Comment (moraaar) [inline Gems/Atom/RHI/Vulkan/Code/Source/RHI/Allocator.h:18]: nit: An API comment `//!` with a description of the class purpose is always helpful.

**Review by moraaar — APPROVED** [review]
LGTM. The gem [OpenXRVk](https://github.com/o3de/o3de-extras/tree/development/Gems/OpenXRVk) is using Vulkan too, would that need to be updated as well?

If the answer is yes and we want to use the same memory manager then `VkSystemAllocator::Get()` would have to be somewhere under `Gems/Atom/RHI/Vulkan/Code/Include/...` so the gem can reach it. Keep this in mind.

File: Gems/Atom/RHI/Vulkan/Code/Source/RHI/Allocator.h (hunk 2 above)
Comment (moudgils) [inline Gems/Atom/RHI/Vulkan/Code/Source/RHI/Allocator.h:18]: Add a comment explaining the purpose of this class.

**Review by moudgils — CHANGES_REQUESTED** [review]
I think you will need to expand ReleaseContainer to take the new allocator. This container is used to destroy Vk objects after 3 frames and at the moment it is passing in null here.

~ReleaseContainer() { m_vkDestroyFunc(m_vkDevice, m_vkObject, nullptr); }

File: Gems/Atom/RHI/Vulkan/Code/Source/RHI/ImageView.cpp [hunk 3]
```diff
@@ -103,7 +104,8 @@ namespace AZ
             createInfo.components = VkComponentMapping{}; // identity mapping
             createInfo.subresourceRange = vkRange;
 
-            const VkResult result = device.GetContext().CreateImageView(device.GetNativeDevice(), &createInfo, nullptr, &m_vkImageView);
+            const VkResult result =
+                device.GetContext().CreateImageView(device.GetNativeDevice(), &createInfo, VkSystemAllocator::Get(), &m_vkImageView);
```
[the remark this answers is not in the document]
Comment (Jackie9527) [reply Gems/Atom/RHI/Vulkan/Code/Source/RHI/ImageView.cpp:108]: The function call of `DestroyImageView` is queued by using `ReleaseContainer`.

File: Gems/Atom/RHI/Vulkan/Code/Source/RHI/QueryPool.cpp [hunk 4]
```diff
@@ -73,7 +74,8 @@ namespace AZ
             createInfo.queryCount = descriptor.m_queriesCount;
             createInfo.pipelineStatistics = ConvertQueryPipelineStatisticMask(descriptor.m_pipelineStatisticsMask);
 
-            auto vkResult = device.GetContext().CreateQueryPool(device.GetNativeDevice(), &createInfo, nullptr, &m_nativeQueryPool);
+            auto vkResult =
+                device.GetContext().CreateQueryPool(device.GetNativeDevice(), &createInfo, VkSystemAllocator::Get(), &m_nativeQueryPool);
```
[the remark this answers is not in the document]
Comment (Jackie9527) [reply Gems/Atom/RHI/Vulkan/Code/Source/RHI/QueryPool.cpp:78]: The function call of `DestroyImageView` is queued by using `ReleaseContainer`.

File: Gems/Atom/RHI/Vulkan/Code/Source/RHI/Allocator.h (hunk 1 above)
Comment (Jackie9527) [reply Gems/Atom/RHI/Vulkan/Code/Source/RHI/Allocator.h]: The repeated `private` has been removed.

File: Gems/Atom/RHI/Vulkan/Code/Source/RHI/Allocator.h (hunk 2 above)
Comment (Jackie9527) [reply Gems/Atom/RHI/Vulkan/Code/Source/RHI/Allocator.h]: done.

Comment (Jackie9527) [reply Gems/Atom/RHI/Vulkan/Code/Source/RHI/Allocator.h]: done.

File: Gems/Atom/RHI/Vulkan/Code/Source/RHI/Allocator.h [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+#include <AzCore/PlatformIncl.h>
+#include <vulkan/vulkan.h>
+
+namespace AZ
+{
+    namespace Vulkan
+    {
+        class VkSystemAllocator 
+        {
+
```
[the remark this answers is not in the document]
Comment (Jackie9527) [reply Gems/Atom/RHI/Vulkan/Code/Source/RHI/Allocator.h]: done

**Comment by Jackie9527** [comment]
> I think you will need to expand ReleaseContainer to take the new allocator. This container is used to destroy Vk objects after 3 frames and at the moment it is passing in null here.
> 
> ~ReleaseContainer() { m_vkDestroyFunc(m_vkDevice, m_vkObject, nullptr); }

done.

**Comment by Jackie9527** [comment]
> LGTM. The gem [OpenXRVk](https://github.com/o3de/o3de-extras/tree/development/Gems/OpenXRVk) is using Vulkan too, would that need to be updated as well?
> 
> If the answer is yes and we want to use the same memory manager then `VkSystemAllocator::Get()` would have to be somewhere under `Gems/Atom/RHI/Vulkan/Code/Include/...` so the gem can reach it. Keep this in mind.

The header file is moved to `Gems/Atom/RHI/Vulkan/Code/Include/Atom/RHI`, and renamed from `Allocator.h` to `VkAllocator.h`.

File: Gems/Atom/RHI/Vulkan/Code/Include/Atom/RHI/VkAllocator.h [hunk 6]
```diff
@@ -0,0 +1,31 @@
+/*
```
Comment (moudgils) [inline Gems/Atom/RHI/Vulkan/Code/Include/Atom/RHI/VkAllocator.h:1]: The current pattern is setup where all the header files actually live within C:\GitHub\o3de\Gems\Atom\RHI\Vulkan\Code\Source\RHI. Please move this file to that directory. Thanks.

Comment (moudgils) [inline Gems/Atom/RHI/Vulkan/Code/Include/Atom/RHI/VkAllocator.h:1]: This header will need to go to RHI.Reflect folder as we pull that into OpenXrVk gem. Also, are you working on a PR to update the OpenXrVk gem as well (to use this allocator)?

Comment (Jackie9527) [reply Gems/Atom/RHI/Vulkan/Code/Include/Atom/RHI/VkAllocator.h]: The header file is moved to `Gems/Atom/RHI/Vulkan/Code/Include/Atom/RHI.Reflect`.
I will see to `OpenXrVk` after #14526 is merged.

File: Gems/Atom/RHI/Vulkan/Code/Source/RHI/VkAllocator.cpp [hunk 7]
```diff
@@ -0,0 +1,61 @@
+/*
```
Comment (moudgils) [inline Gems/Atom/RHI/Vulkan/Code/Source/RHI/VkAllocator.cpp:1]: This file needs to also be relocated to RHI.Reflect folder.

File: Gems/Atom/RHI/Vulkan/Code/atom_rhi_vulkan_private_common_files.cmake [hunk 8]
```diff
@@ -7,6 +7,8 @@
 #
 
 set(FILES
+    Include/Atom/RHI.Reflect/VkAllocator.h
```
Comment (moudgils) [inline Gems/Atom/RHI/Vulkan/Code/atom_rhi_vulkan_private_common_files.cmake:10]: These changes will need to be moved to atom_rhi_vulkan_reflect_common_files

Comment (Jackie9527) [reply Gems/Atom/RHI/Vulkan/Code/atom_rhi_vulkan_private_common_files.cmake]: done

File: Gems/Atom/RHI/Vulkan/Code/Source/RHI/VkAllocator.cpp (hunk 7 above)
Comment (Jackie9527) [reply Gems/Atom/RHI/Vulkan/Code/Source/RHI/VkAllocator.cpp]: done

**Review by moudgils — APPROVED** [review]

**Comment by Jackie9527** [comment]
> LGTM. The gem [OpenXRVk](https://github.com/o3de/o3de-extras/tree/development/Gems/OpenXRVk) is using Vulkan too, would that need to be updated as well?
> 
> If the answer is yes and we want to use the same memory manager then `VkSystemAllocator::Get()` would have to be somewhere under `Gems/Atom/RHI/Vulkan/Code/Include/...` so the gem can reach it. Keep this in mind.

please review [190](https://github.com/o3de/o3de-extras/pull/190), thx.

**Pull request merged**